File: api/mdx.go
package api

import (
	"regexp"
	"strings"
)

// FenceLineRegex matches fence marker lines (``` or ~~~) with optional leading whitespace.
// Use with FindAllStringIndex to find fence positions for pairing.
var FenceLineRegex = regexp.MustCompile(`(?m)^\s*(?:` + "```" + `|~~~)`)

// fencedBlockRegex matches complete fenced code blocks (opening + content + closing).
// Used for stripping entire blocks from content.
var fencedBlockRegex = regexp.MustCompile(`(?m)^\s*(?:` + "```" + `|~~~).*\n(?:.*\n)*?^\s*(?:` + "```" + `|~~~)\s*$`)

// FindFencedCodeBlockRanges finds all fenced code block ranges in the content.
// Returns a slice of [start, end] pairs representing positions inside code blocks.
// This is used to skip components that appear as documentation examples inside code blocks.
func FindFencedCodeBlockRanges(content string) [][2]int {
	var ranges [][2]int

	// Find all fence markers (lines starting with ``` or ~~~)
	// In markdown, fences alternate: open, close, open, close...
	matches := FenceLineRegex.FindAllStringIndex(content, -1)

	// Pair up consecutive fences: [0,1], [2,3], [4,5], etc.
	for i := 0; i+1 < len(matches); i += 2 {
		openStart := matches[i][0]
		closeStart := matches[i+1][0]
		// Find end of closing fence line
		closeEnd := len(content)
		if nl := strings.IndexByte(content[closeStart:], '\n'); nl != -1 {
			closeEnd = closeStart + nl + 1
		}
		ranges = append(ranges, [2]int{openStart, closeEnd})
	}

	return ranges
}

// IsInsideFencedCodeBlock checks if a position is inside any fenced code block.
func IsInsideFencedCodeBlock(position int, codeBlockRanges [][2]int) bool {
	for _, r := range codeBlockRanges {
		if position >= r[0] && position < r[1] {
			return true
		}
	}
	return false
}

// StripFencedCodeBlocks removes fenced code block content from MDX content.
// This prevents false positives when parsing blocks that appear inside documentation examples.
// Matches both ``` and ~~~ fences, with optional leading whitespace.
func StripFencedCodeBlocks(content string) string {
	return fencedBlockRegex.ReplaceAllString(content, "")
}

File: src/mdx.ts
/**
 * Matches fence marker lines (``` or ~~~ runs of 3+) and captures the run and
 * the rest of the line. Any indent is allowed because MDX turns off indented
 * code; `[ \t]*` (not `\s*`) keeps a match from starting on an earlier line.
 */
const FENCE_LINE_REGEX = /^[ \t]*(`{3,}|~{3,})(.*)$/gm

/**
 * Finds all fenced code block ranges as [start, end] position pairs.
 * Used to prevent parsing documentation examples inside code fences.
 *
 * Fences pair the way CommonMark pairs them: a block closes only on a run of
 * the same character that is at least as long as the opener and has nothing
 * after it. A nested ```mdx example inside a ```` fence, or an indented
 * ```yaml line inside a ```mdx fence, is content rather than a closer. An
 * unclosed fence runs to the end of the content.
 */
export function findFencedCodeBlockRanges(content: string): Array<[number, number]> {
  const ranges: Array<[number, number]> = []
  let open: { start: number; char: string; length: number } | null = null

  FENCE_LINE_REGEX.lastIndex = 0
  let m: RegExpExecArray | null
  while ((m = FENCE_LINE_REGEX.exec(content)) !== null) {
    const run = m[1]!
    const rest = m[2]!
    const char = run[0]!

    if (!open) {
      // A backtick opener's info string can't contain backticks (```x``` is inline code)
      if (char === "`" && rest.includes("`")) continue
      open = { start: m.index, char, length: run.length }
    } else if (char === open.char && run.length >= open.length && rest.trim() === "") {
      ranges.push([open.start, m.index + m[0].length])
      open = null
    }
  }

  if (open) ranges.push([open.start, content.length])
  return ranges
}

/**
 * Returns true if `position` falls inside any fenced code block.
 */
export function isInsideFencedCodeBlock(position: number, ranges: Array<[number, number]>): boolean {
  return ranges.some(([start, end]) => position >= start && position <= end)
}
